File: jianzhi_offer/No.07 重建二叉树/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        return self.build(preorder, inorder, 0, len(preorder), 0, len(inorder))
    def build(self,preorder,inorder,pre_start,pre_end,in_start,in_end):
        if in_start==in_end:return None
        root=TreeNode(preorder[pre_start])
        root_idx=inorder.index(preorder[pre_start])
        root.left=self.build(preorder, inorder, pre_start+1, pre_start+1+(root_idx-in_start), in_start, root_idx)
        root.right=self.build(preorder, inorder, pre_start+1+(root_idx-in_start),pre_end , root_idx+1,in_end)
        return root

File: jianzhi_offer/No.07 重建二叉树/solution.cpp
class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        return build(preorder.begin(),preorder.end(),inorder.begin(),inorder.end());
    }
    TreeNode* build(vector<int>::iterator prestart,vector<int>::iterator preend,vector<int>::iterator instart,vector<int>::iterator inend){
        if(instart==inend) return NULL;
        TreeNode* root=new TreeNode(*prestart);//指向根节点的指针
        auto root_idx=find(instart,inend,*prestart);//查找根节点在中序的位置
        root->left=build(prestart+1,prestart+1+(root_idx-instart),instart,root_idx);//建立左子树
        root->right=build(prestart+1+(root_idx-instart),preend,root_idx+1,inend);//建立右子树
        return root;
    }
};